control/pure_pursuit/src/pure_pursuit/pure_pursuit_lateral_controller.cpp

```diff
@@ -323,12 +394,69 @@ boost::optional<PpOutput> PurePursuitLateralController::calcTargetCurvature(
   }
 
   // Calculate target point for velocity/acceleration
-  const auto target_point = calcTargetPoint(pose);
-  if (!target_point) {
+
+  const auto closest_idx_result =
+    motion_utils::findNearestIndex(*output_tp_array_, pose, 3.0, M_PI_4);
+  if (!closest_idx_result) {
+    RCLCPP_ERROR(node_->get_logger(), "cannot find closest waypoint");
+    return {};
+  }
+
+  const double && target_vel =
+    trajectory_resampled_->points.at(*closest_idx_result).longitudinal_velocity_mps;
+
+  // calculate the lateral error
+
+  geometry_msgs::msg::Point vec_end;
+  geometry_msgs::msg::Point vec_start;
+
+  if (*closest_idx_result == output_tp_array_->size() - 1) {
+    vec_end = output_tp_array_->at(*closest_idx_result).pose.position;
+    vec_start = output_tp_array_->at(*closest_idx_result - 1).pose.position;
+  } else {
+    vec_start = output_tp_array_->at(*closest_idx_result).pose.position;
+    vec_end = output_tp_array_->at(*closest_idx_result + 1).pose.position;
+  }
+
+  Eigen::Vector3d vec_a(
+    (vec_end.x - vec_start.x), (vec_end.y - vec_start.y), (vec_end.z - vec_start.z));
+
+  if (vec_a.norm() < threshold) {
+    RCLCPP_ERROR(node_->get_logger(), "Trajectory point interval is almost 0.");
     return {};
   }
 
-  const double target_vel = target_point->longitudinal_velocity_mps;
+  const double lateral_error =
+    planning_utils::calcLateralError2D(vec_start, vec_end, pose.position);
+
+  // Calculate current curvature
+  const size_t idx_dist = static_cast<size_t>(
+    std::max(static_cast<int>((param_.curvature_calculation_distance) / param_.resampling_ds), 1));
+
+  // Find the points in trajectory to calculate curvature
+  size_t next_idx = trajectory_resampled_->points.size() - 1;
+  size_t prev_idx = 0;
+
+  if (static_cast<size_t>(*closest_idx_result) >= idx_dist) {
+    prev_idx = *closest_idx_result - idx_dist;
+  }
+  if (trajectory_resampled_->points.size() - 1 >= *closest_idx_result + idx_dist) {
+    next_idx = *closest_idx_result + idx_dist;
+  }
+
+  // Calculate curvature assuming the trajectory points interval is constant
+  double current_curvature = 0.0;
+
+  try {
+    current_curvature = calcCurvature(
+      trajectory_resampled_->points.at(prev_idx),
+      trajectory_resampled_->points.at(*closest_idx_result),
+      trajectory_resampled_->points.at(next_idx));
+  } catch (std::exception const & e) {
+    // ...code that handles the error...
+    RCLCPP_WARN(rclcpp::get_logger("pure_pursuit"), "%s", e.what());
+    current_curvature = 0.0;
+  }
```

Comment:
Would you make another function for the trajectory curvature computation? It can simplify the operation flow of this function.